### PR TITLE
Added retention feature to table manager to automatically remove olde…

### PR DESCRIPTION
**Pull request by rsteneteg**
…r tables

Signed-off-by: Roger Steneteg <rsteneteg@ea.com>

### CODE REVIEW

**Comment by rsteneteg** [comment]
this fixes https://github.com/cortexproject/cortex/issues/1132

File: pkg/chunk/schema_config.go [hunk 1]
```diff
@@ -299,15 +299,17 @@ func (cfg *PeriodConfig) dailyBuckets(from, through model.Time, userID string) [
...
+	Period    time.Duration `yaml:"period,omitempty"`
+	Retention int64         `yaml:"retention,omitempty"`
+	Tags      Tags          `yaml:"tags,omitempty"`
 }
 
 // RegisterFlags adds the flags required to config this to the given FlagSet.
 func (cfg *PeriodicTableConfig) RegisterFlags(argPrefix, tablePrefix string, f *flag.FlagSet) {
 	f.StringVar(&cfg.Prefix, argPrefix+".prefix", tablePrefix, "DynamoDB table prefix for period tables.")
 	f.DurationVar(&cfg.Period, argPrefix+".period", 7*24*time.Hour, "DynamoDB table period.")
+	f.Int64Var(&cfg.Retention, argPrefix+".retention", 0, "Number of tables to keep, (0 disables retention).")
```
Comment (tomwilkie) [inline pkg/chunk/schema_config.go]: Would it be possible to set retention duration?  Having this as a number means I have to know "how long" a table is, which might change.

**Comment by tomwilkie** [comment]
Nice, thanks @rsteneteg!  This is something I was just thinking about too - you saved me the time.  Implementation looks really good; only one question.  

I'll also ask @bboreham and @khaines for reviews here, as its quite sensitive code.  And I'd like to deploy to our test env.  Do you guys have this running now?

File: pkg/chunk/schema_config.go (hunk 1 above)
Comment (rsteneteg) [reply pkg/chunk/schema_config.go]: I suppose we could have retention as a duration and get the number of tables to keep by dividing it by the table period. That way you could change the period without needing to update the retention. Is that what you was referring to with that table length may change?

Comment (tomwilkie) [reply pkg/chunk/schema_config.go]: Eggsactly

Comment (rsteneteg) [reply pkg/chunk/schema_config.go]: I like it, I will fix that

Comment (rsteneteg) [reply pkg/chunk/schema_config.go]: It is now a duration, thanks for suggesting that change

**Comment by rsteneteg** [comment]
@tomwilkie as far as testing it goes, I have tested it on my machine against DynamoDB, I am working on getting the code running in our testing pipeline.

**Comment by rsteneteg** [comment]
Even if `retention` is set to `0`, this feature would still delete tables that is older than the configs `from` date. If we want any deletes to be disabled by default we would need to change the default of the `table-manager.retention-deletes-disabled` parameter to be true, or rather rename it to `retention-deletes-enabled` with default false to avoid having a negative bool flag.

**Comment by bboreham** [comment]
I would expect if there is no `Retention` setting then no deletions. 

As it happens I have set `from` to last Wednesday in order to use on-demand pricing: I don’t want every table deleted.

**Comment by rsteneteg** [comment]
Made more sense to have a single retention parameter on table manager instead in each config in the schema. Hopefully there is no use case where you would have wanted different retention per chema config

File: pkg/chunk/table_manager.go [hunk 2]
```diff
@@ -274,22 +286,39 @@ func (m *TableManager) calculateExpectedTables() []TableDesc {
...
 			toCreate = append(toCreate, descriptions[i])
 			i++
 		} else if descriptions[i].Name > existingTables[j] {
-			// existingTables[j].name isn't in descriptions, can ignore
+			// existingTables[j].name isn't in descriptions, and can be removed
+			if m.cfg.RetentionPeriod > 0 {
+				isMatch := false
+				for tblPrefix := range tablePrefixes {
+					if strings.HasPrefix(existingTables[j], tblPrefix) {
+						isMatch = true
```
[the remark this answers is not in the document]
Comment (rsteneteg) [reply pkg/chunk/table_manager.go]: Fixed it, thanks you

File: pkg/chunk/schema_config.go [hunk 3]
```diff
@@ -299,15 +299,17 @@ func (cfg *PeriodConfig) dailyBuckets(from, through model.Time, userID string) [
...
+	Period    time.Duration `yaml:"period,omitempty"`
+	Retention time.Duration `yaml:"retention,omitempty"`
+	Tags      Tags          `yaml:"tags,omitempty"`
 }
 
 // RegisterFlags adds the flags required to config this to the given FlagSet.
 func (cfg *PeriodicTableConfig) RegisterFlags(argPrefix, tablePrefix string, f *flag.FlagSet) {
 	f.StringVar(&cfg.Prefix, argPrefix+".prefix", tablePrefix, "DynamoDB table prefix for period tables.")
 	f.DurationVar(&cfg.Period, argPrefix+".period", 7*24*time.Hour, "DynamoDB table period.")
+	f.DurationVar(&cfg.Retention, argPrefix+".retention", 0, "Tables passed this retention period is deleted, (0 disables retention).")
```
[the remark this answers is not in the document]
Comment (rsteneteg) [reply pkg/chunk/schema_config.go]: I think period keeps it more consistent with other duration parameters.

I agree that your suggested parameter description is better, I also changed the "dry-run" parameter enabled by default, as an extra security precaution before deleting tables. 
Thank you, PR updated

File: pkg/chunk/aws/dynamodb_table_client.go [hunk 4]
```diff
@@ -189,6 +189,24 @@ func (d dynamoTableClient) CreateTable(ctx context.Context, desc chunk.TableDesc
 	return nil
 }
 
+func (d dynamoTableClient) DeleteTable(ctx context.Context, name string) error {
+	if err := d.backoffAndRetry(ctx, func(ctx context.Context) error {
```
Comment (bboreham) [inline pkg/chunk/aws/dynamodb_table_client.go]: Seeing this prompted me to file #1152. (just a comment; no action required in this PR)

File: pkg/chunk/table_manager.go [hunk 5]
```diff
@@ -72,6 +78,8 @@ type ProvisionConfig struct {
 // RegisterFlags adds the flags required to config this to the given FlagSet.
 func (cfg *TableManagerConfig) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.ThroughputUpdatesDisabled, "table-manager.throughput-updates-disabled", false, "If true, disable all changes to DB capacity")
+	f.BoolVar(&cfg.RetentionDeletesEnabled, "table-manager.retention-deletes-enabled", false, "If true, enables retention deletes of DB tables")
```
Comment (bboreham) [inline pkg/chunk/table_manager.go]: Why do we need both (a boolean flag to enable) and (0 duration to disable) ?

Comment (rsteneteg) [reply pkg/chunk/table_manager.go]: The thought behind that was that you could set retention and keep deletes off to verify in the logs that it wants to delete the correct tables, before actually turning on the actual deletes.

**Pull request merged**